add cli describer for oauthtokens

juanvallejo opened the pull request:
Adds a command line describer for OAuth tokens:
`oc describe oauthaccesstoken <token-name>`

```
Name:           <token-name>
Namespace:      <none>
Created:        2 hours ago
Labels:         <none>
Annotations:    <none>
Scopes:         []
Expires In:     24 hours
User Name:      test
User UID:       322b236b-22b9-11e6-b307-080027242396
Refresh Token:  <none>
Client Name:    openshift-challenging-client
```

fixes https://github.com/openshift/origin/issues/9012

deads2k commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -418,6 +420,36 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
+		formatMeta(out, oAuthAccessToken.ObjectMeta)
+		formatString(out, "Client Name", oAuthAccessToken.ClientName)
+		formatString(out, "Expires In", oAuthAccessToken.ExpiresIn)
+		formatString(out, "Redirect URI", oAuthAccessToken.RedirectURI)
```
i don't think this value is very useful.  The redirect already happened and someone doing a post-mortem on a token will have the json.

deads2k commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -418,6 +420,36 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
+	return tabbedString(func(out *tabwriter.Writer) error {
+		formatMeta(out, oAuthAccessToken.ObjectMeta)
+		formatString(out, "Client Name", oAuthAccessToken.ClientName)
+		formatString(out, "Expires In", oAuthAccessToken.ExpiresIn)
```
Preprocess this value so that its a duration from now or a datetime when its done.  Maybe either depending on distance?  "5 minutes" or "2 days", but "october 6, 2016". 

@fabianofranz do we have a time interpretting helper for this?

deads2k commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -418,6 +420,36 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
+
+	return tabbedString(func(out *tabwriter.Writer) error {
+		formatMeta(out, oAuthAccessToken.ObjectMeta)
+		formatString(out, "Client Name", oAuthAccessToken.ClientName)
```
less useful, move it down.

deads2k commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -418,6 +420,36 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
+		formatString(out, "Redirect URI", oAuthAccessToken.RedirectURI)
+		formatString(out, "User Name", oAuthAccessToken.UserName)
+		formatString(out, "User UID", oAuthAccessToken.UserUID)
+		formatString(out, "Authorize Token", oAuthAccessToken.AuthorizeToken)
```
Not important enough to display.

deads2k commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -418,6 +420,36 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
+		formatString(out, "Authorize Token", oAuthAccessToken.AuthorizeToken)
+		formatString(out, "Refresh Token", oAuthAccessToken.RefreshToken)
+		if len(name) > 0 {
+			formatString(out, "Token Name", name)
```
already shown by `formatMeta`, right?

deads2k commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -418,6 +420,36 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
+		if len(name) > 0 {
+			formatString(out, "Token Name", name)
+		}
+		formatString(out, "Scopes", oAuthAccessToken.Scopes)
```
These are very important.  I'd put them right after objectmeta.

deads2k commented:
Looks good, the output just needs a little swizzling.

juanvallejo commented:
Updated output:

`oc describe oauthaccesstoken <token-name>`

```
Name:           <token-name>
Namespace:      <none>
Created:        2 hours ago
Labels:         <none>
Annotations:    <none>
Scopes:         []
Expires In:     24 hours
User Name:      test
User UID:       322b236b-22b9-11e6-b307-080027242396
Refresh Token:  <none>
Client Name:    openshift-challenging-client
```

liggitt commented:
expires in should either be an absolute date, or relative to the current time

deads2k commented:
> expires in should either be an absolute date, or relative to the current time

Do you happen to know the answer to my question; https://github.com/openshift/origin/pull/9032#discussion_r64733345 ?

deads2k commented:
@juanvallejo found it . `FormatRelativeTime`

juanvallejo commented:
Updated expiration time to absolute date (helper functions that we have fail to correctly handle a time difference longer than 24h so not using FormatRelativeTime for now, I will submit another pull request addressing this).

Updated console output:

`oc describe oauthaccesstoken <token-name>`

```
Name:           mHVzxUncDlnOQZjK7qEXuoTYYzGnWVlrwNOmFJlTmi4
Namespace:      <none>
Created:        6 hours ago
Labels:         <none>
Annotations:    <none>
Scopes:         []
Expires:        May 27, 14:37 +0000 UTC
User Name:      test
User UID:       322b236b-22b9-11e6-b307-080027242396
Refresh Token:  <none>
Client Name:    openshift-challenging-client
```

juanvallejo commented:
Updated console output:

`oc describe oauthaccesstoken <token-name>`

```
Name:           <token-name>
Namespace:      <none>
Created:        11 hours ago
Labels:         <none>
Annotations:    <none>
Scopes:         []
Expires In:     12 hours
User Name:      test
User UID:       322b236b-22b9-11e6-b307-080027242396
Refresh Token:  <none>
Client Name:    openshift-challenging-client
```

deads2k commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -456,6 +458,34 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
+		formatString(out, "Expires In", formatToHumanDuration(timeExpired.Sub(time.Now())))
+		formatString(out, "User Name", oAuthAccessToken.UserName)
+		formatString(out, "User UID", oAuthAccessToken.UserUID)
+		formatString(out, "Refresh Token", oAuthAccessToken.RefreshToken)
```
We don't provide these yet.  Remove this line.

deads2k commented on test/cmd/basicresources.sh:
```diff
@@ -250,6 +250,11 @@ os::cmd::expect_success_and_text "oc get route route-with-set-port --template='{
 os::test::junit::declare_suite_end
 
+# Test OAuth access token describer
+os::cmd::expect_success 'oc create -f test/fixtures/oauthaccesstoken.yaml'
```
Does this test pass?  That access token looks invalid.

deads2k commented:
Can you show me an example 6 months in the future?

deads2k commented on pkg/cmd/cli/describe/helpers.go:
```diff
@@ -118,6 +118,16 @@ var timeNowFn = func() time.Time {
 	return time.Now()
 }
 
+func TimeNowFn() time.Time {
```
Why are you exposing this?

juanvallejo commented:
@deads2k Output on token with expiration date 6months in the future:

`oc describe oauthaccesstoken DYGZDLucARCPIfUeKPhsgPfn0WBLR_9KdeREH0c9ioc`

```
Name:           DYGZDLucARCPIfUeKPhsgPfn0WBLR_9KdeREH0c9ioc
Namespace:      <none>
Created:        13 minutes ago
Labels:         <none>
Annotations:    <none>
Scopes:         []
Expires In:     5 months
User Name:      test
User UID:       322b236b-22b9-11e6-b307-080027242396
Client Name:    openshift-challenging-client
```

deads2k commented:
> @deads2k Output on token with expiration date 6months in the future:

Ok, go ahead and squash down

deads2k commented on test/fixtures/oauthaccesstoken.yaml:
```diff
@@ -0,0 +1,14 @@
+  creationTimestamp: 2016-05-27T15:06:11Z
+  deletionTimestamp: 2016-11-28T15:06:11Z
+  name: DYGZDLucARCPIfUeKPhsgPfn0WBLR_9KdeREH0c9iod
+  resourceVersion: "5338"
```
remove creation,deleteion,resourceversion,selflink,uid

deads2k commented on test/fixtures/oauthaccesstoken.yaml:
```diff
@@ -0,0 +1,14 @@
+  resourceVersion: "5338"
+  selfLink: /oapi/v1/oauthaccesstokens/DYGZDLucARCPIfUeKPhsgPfn0WBLR_9KdeREH0c9iod
+  uid: 8c3a2a8c-241c-11e6-8919-080027242396
+redirectURI: https://10.0.2.15:8443/oauth/token/implicit
```
make this ip either localhost or a stub

deads2k commented on test/fixtures/oauthaccesstoken.yaml:
```diff
@@ -0,0 +1,14 @@
+  uid: 8c3a2a8c-241c-11e6-8919-080027242396
+redirectURI: https://10.0.2.15:8443/oauth/token/implicit
+userName: test
+userUID: 322b236b-22b9-11e6-b307-080027242396
```
this user UID will be invalid, if the test works, I don't care I guess.

juanvallejo replied:
yep, test working fine with it

deads2k commented on pkg/client/oauthaccesstoken.go:
```diff
@@ -25,6 +26,13 @@ func newOAuthAccessTokens(c *Client) *oauthAccessTokenInterface {
 }
 
+// Get returns information about a particular image and error if one occurs.
+func (c *oauthAccessTokenInterface) Get(name string) (result *oauthapi.OAuthAccessToken, err error) {
```
Why add this?  I don't object, but I'm pretty sure you didn't need to.

juanvallejo replied:
I use it to retrieve token contents when outputting token info: https://github.com/juanvallejo/origin/blob/jvallejo_add-cli-oauth-describer/pkg/cmd/cli/describe/describer.go#L468

juanvallejo commented:
@deads2k Okay, revised `test/fixtures/oauthaccesstoken.yaml ` and squashed down

deads2k commented:
lgtm [merge]
